src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala

```diff
@@ -323,16 +318,61 @@ class RideHailManager(
+        .getVehicleAdjustment(beamServices)
+        .sampleRideHailVehicleTypes(
+          numVehicles = numRideHailAgents.toInt,
+          vehicleCategory = VehicleCategory.Car
+        )
+      var activityEndTimes: ArrayBuffer[Int] = new ArrayBuffer[Int]()
+      scenario.getPopulation.getPersons.asScala.map(
+        _._2.getSelectedPlan.getPlanElements.asScala
+          .filter(_.isInstanceOf[Activity])
+          .map(_.asInstanceOf[Activity].getEndTime.toInt)
+          .filter(_ > 0)
+          .map(activityEndTimes += _)
+      )
       val fleetData: ArrayBuffer[RideHailFleetInitializer.RideHailAgentInputData] = new ArrayBuffer(persons.size)
-      persons.foreach { person =>
-        try {
-          val rideInitialLocation: Location = getRideInitLocation(person)
-          fleetData += createRideHailVehicleAndAgent(person.getId.toString, rideInitialLocation, None, None)
-        } catch {
-          case ex: Throwable =>
-            log.error(ex, s"Could not createRideHailVehicleAndAgent: ${ex.getMessage}")
-            throw ex
-        }
+      persons.zipWithIndex.foreach {
+        case (person, idx) =>
+          try {
+            val vehicleType = vehicleTypes(idx)
+            if (beamServices.beamConfig.beam.agentsim.agents.rideHail.refuelThresholdInMeters >=
+                  (vehicleType.primaryFuelCapacityInJoule / vehicleType.primaryFuelConsumptionInJoulePerMeter) * 0.8) {
+//              throw new RuntimeException(
+//                "Ride Hail refuel threshold is higher than state of energy of a vehicle fueled by a DC fast charger. This will cause an infinite loop"
+//              )
+              println("OH NO, THIS SHOULDN'T HAPPEN")
+            }
+            val rideInitialLocation: Location = getRideInitLocation(person)
+            if (vehicleType.automationLevel < 4) {
+              val shiftDuration = math.round(math.exp(rand.nextGaussian() * 0.67 + 0.60) * 3600)
+              val shiftMidPointTime = activityEndTimes(rand.nextInt(activityEndTimes.length))
+              val shiftStartTime = max(shiftMidPointTime - (shiftDuration / 2).toInt, 10)
+              val shiftEndTime = min(shiftMidPointTime + (shiftDuration / 2).toInt, 24 * 3600)
+              val shiftString = convertToShiftString(ArrayBuffer(shiftStartTime), ArrayBuffer(shiftEndTime))
+              fleetData += createRideHailVehicleAndAgent(
+                person.getId.toString,
+                vehicleType,
+                rideInitialLocation,
+                shiftString,
+                None
+              )
+            } else {
+              val shiftString = convertToShiftString(ArrayBuffer(0), ArrayBuffer(24 * 3600))
+              fleetData += createRideHailVehicleAndAgent(
+                person.getId.toString,
+                vehicleType,
+                rideInitialLocation,
+                shiftString,
+                None
+              )
+            }
+          } catch {
+            case ex: Throwable =>
```

Comment:
There is no point in catching and rethrowing. Let the top level handle the catching and logging if that's all you are doing.